contracts/exchange/contracts/src/MixinSignatureValidator.sol

```diff
@@ -305,6 +350,53 @@ contract MixinSignatureValidator is
         ));
     }
 
+    /// @dev Verifies signature using logic defined by an EIP1271 Wallet contract.
+    /// @param hash Any 32 byte hash.
+    /// @param walletAddress Address that should have signed the given hash
+    ///                      and defines its own signature verification method.
+    /// @param signature Proof that the hash has been signed by signer.
+    /// @return True if the signature is validated by the Walidator.
+    function _validateHashWithEIP1271Wallet(
+        bytes32 hash,
+        address walletAddress,
+        bytes memory signature
+    )
+        private
+        view
+        returns (bool isValid)
+    {
+        uint256 signatureLength = signature.length;
+        // Shave the signature type off the signature.
+        assembly {
+            mstore(signature, sub(signatureLength, 1))
+        }
+        // Encode the call data.
+        bytes memory data = new bytes(32);
+        data.writeBytes32(0, hash);
+        bytes memory callData = abi.encodeWithSelector(
+            IEIP1271Wallet(walletAddress).isValidSignature.selector,
+            data,
+            signature
+        );
+        // Restore the full signature.
+        assembly {
+            mstore(signature, signatureLength)
+        }
+        // Static call the verification function.
+        (bool didSucceed, bytes memory returnData) = walletAddress.staticcall(callData);
+        // Return data should be the `EIP1271_MAGIC_VALUE`.
+        if (didSucceed && returnData.length == 32) {
+            return bytes4(returnData.readBytes32(0)) == EIP1271_MAGIC_VALUE;
```

Comment:
Why not use `readBytes4` here?

Reply:
Just to be extra sure they meant to return the magic bytes and not something that just coincidentally starts with the magic bytes, I suppose.

Reply:
Doh, you guys are totes right. Put the cast on the wrong side.